app/push_notification.go

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import (
+	"hash/fnv"
+	"net/http"
+	"strings"
+
+	l4g "github.com/alecthomas/log4go"
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/utils"
+	"github.com/nicksnyder/go-i18n/i18n"
+)
+
+const PUSH_NOTIFICATION_HUB_WORKERS = 1000
+const PUSH_NOTIFICATIONS_HUB_BUFFER_PER_WORKER = 50
```

Comment:
We can move this two constants in a settings but only would have effect on restart.

Reply:
The limit here would be a single user receiving 50 push notifications, in the time you need to send one. This means that it will block any gorouting that needs to send new notifications to this user until this buffer free some space. Probably that is the reason to move that to a setting, and allow users to define their own policy about the push notifications hub.